### PR TITLE
ARROW-7671: [Python][Dataset] Add bindings for the DatasetFactory

### PR DESCRIPTION
**Pull request by kszucs**
To support constructing dataset from multiple sources.

### CODE REVIEW

File: python/pyarrow/dataset.py [hunk 1]
```diff
@@ -203,6 +204,15 @@ def _ensure_partitioning(scheme):
     return scheme
 
 
+def _ensure_format(obj):
+    if isinstance(obj, FileFormat):
+        return obj
+    elif obj == "parquet":
+        return ParquetFileFormat()
+    elif not isinstance(obj, FileFormat):
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py]:
```suggestion
    else:
```

you already checked above for FileFormat

File: python/pyarrow/dataset.py [hunk 2]
```diff
@@ -219,59 +229,48 @@ def source(path_or_paths, filesystem=None, partitioning=None,
...
     elif isinstance(partitioning, Partitioning):
         options.partitioning = partitioning
 
-    discovery = FileSystemSourceFactory(
-        filesystem, paths_or_selector, format, options)
-
-    # TODO return Source if a specific schema was passed?
-
-    # need to return SourceFactory since `dataset` might need to
-    # finish the factory with a unified schema
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py]: Can you leave those comments ?

File: python/pyarrow/dataset.py [hunk 3]
```diff
@@ -219,59 +229,48 @@ def source(path_or_paths, filesystem=None, partitioning=None,
...
             raise ValueError(
                 "When passing a Source(Factory), you cannot pass any "
-                "additional arguments")
+                "additional arguments"
+            )
+        return src
+    elif isinstance(src, Source):
+        raise TypeError(
+            "Source objects are currently not supported, only SourceFactory "
+            "instances. Use the ds.source() function to create such objects."
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py]:
```suggestion
            "instances. Use the source() function to create such objects."
```

(or in full, but `ds` is not necessarily known to the user)

File: python/pyarrow/tests/test_dataset.py [hunk 4]
```diff
@@ -583,3 +665,51 @@ def test_filter_implicit_cast(tempdir):
...
+    src2 = ds.source('/schema', filesystem=multisourcefs, format='parquet',
+                     partitioning=['week', 'color'])
+    src3 = ds.source('/hive', filesystem=multisourcefs, format='parquet',
+                     partitioning='hive')
+
+    assembled = ds.dataset([src1, src2, src3])
+    assert isinstance(assembled, ds.Dataset)
+
+    expected_schema = pa.schema([
+        pa.field('date', pa.date32()),
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py]: btw, you don't need the pa.field(), just a tuple is fine as well

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

Just few tiny comments

File: python/pyarrow/dataset.py (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/dataset.py]: Updated.

File: python/pyarrow/dataset.py (hunk 2 above)
Comment (kszucs) [reply python/pyarrow/dataset.py]: Sure, updated.

File: python/pyarrow/dataset.py (hunk 3 above)
Comment (kszucs) [reply python/pyarrow/dataset.py]: Fixed.

File: python/pyarrow/tests/test_dataset.py (hunk 4 above)
Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py]: I'm just used to it, but updated to the less verbose syntax.

**Review by kszucs — COMMENTED** [review]
+1

**Pull request closed without merging**